src/NHibernate/Linq/LinqExtensionMethods.cs

```diff
@@ -8,11 +8,1520 @@
+		}
+
+		/// <summary>
+		/// Computes the average of a sequence of nullable <see cref="T:System.Decimal"/> values that is obtained by invoking a projection function on each element of the input sequence.
+		/// </summary>
+		/// <param name="source">A sequence of values to calculate the average of.</param>
+		/// <param name="selector">A projection function to apply to each element.</param>
+		/// <param name="cancellationToken">A cancellation token that can be used to cancel the work.</param>
+		/// <typeparam name="TSource">The type of the elements of <paramref name="source"/>.</typeparam>
+		/// <returns>
+		/// The average of the sequence of values, or null if the <paramref name="source"/> sequence is empty or contains only null values.
+		/// </returns>
+		/// <exception cref="T:System.ArgumentNullException"><paramref name="source"/> or <paramref name="selector"/> is null.</exception>
+		public static Task<decimal?> AverageAsync<TSource>(this IQueryable<TSource> source, Expression<Func<TSource, decimal?>> selector, CancellationToken cancellationToken = default(CancellationToken))
+		{
+			if (source == null)
+			{
+				throw new ArgumentNullException(nameof(source));
+			}
+			if (selector == null)
+			{
+				throw new ArgumentNullException(nameof(selector));
+			}
+			var provider = (INhQueryProvider)source.Provider;
+			var methodInfo = ReflectionCache.QueryableMethods.AverageWithSelectorOfNullableDecimalDefinition.MakeGenericMethod(typeof(TSource));
+			var expression = new[] { source.Expression, Expression.Quote(selector) };
+			return provider.ExecuteAsync<decimal?>(Expression.Call(null, methodInfo, expression), cancellationToken);
+		}
+
+		#endregion
+
+		#region MinAsync
+
+		/// <summary>
+		/// Returns the minimum value of a generic <see cref="T:System.Linq.IQueryable`1"/>.
+		/// </summary>
+		/// <param name="source">A sequence of values to determine the minimum of.</param>
+		/// <param name="cancellationToken">A cancellation token that can be used to cancel the work.</param>
+		/// <typeparam name="TSource">The type of the elements of <paramref name="source"/>.</typeparam>
+		/// <returns>
+		/// The minimum value in the sequence.
+		/// </returns>
+		/// <exception cref="T:System.ArgumentNullException"><paramref name="source"/> is null.</exception>
+		public static Task<TSource> MinAsync<TSource>(this IQueryable<TSource> source, CancellationToken cancellationToken = default(CancellationToken))
+		{
+			if (source == null)
+			{
+				throw new ArgumentNullException(nameof(source));
+			}
+			var provider = (INhQueryProvider)source.Provider;
+			var methodInfo = ReflectionCache.QueryableMethods.MinDefinition.MakeGenericMethod(typeof(TSource));
+			var expression = new[] { source.Expression };
+			return provider.ExecuteAsync<TSource>(Expression.Call(null, methodInfo, expression), cancellationToken);
+		}
+
+		/// <summary>
+		/// Invokes a projection func
```

Comment:
All lines currently missing the `ConfigureAwait(false)`:
![missingconfigureawaits](https://cloud.githubusercontent.com/assets/5311349/26747841/e04521e4-47f8-11e7-8c61-713951a1a25c.jpg)